controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
-	if err := r.ReconcileNormal(rsCtx); err != nil {
+	result, err := r.ReconcileNormal(rsCtx)
+	if err != nil {
 		rsCtx.Logger.Error(err, "Failed to reconcile VirtualMachineReplicaSet")
 		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{}, nil
+	return result, nil
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet Deletion")
+
+	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
+		defer func() {
+			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", nil, false)
+		}()
+
+		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	}
+
+	return nil
+}
+
+// shouldExcludeVirtualMachine returns true if the VirtualMachine should be
+// filtered out, false otherwise.
+func shouldExcludeVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet, vm *vmopv1.VirtualMachine) bool {
+	if metav1.GetControllerOf(vm) != nil && !metav1.IsControlledBy(vm, rs) {
+		return true
+	}
+
+	return false
 }
 
-func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+// adoptOrphan sets the VirtualMachineReplicaSet as a controller OwnerReference
+// to the VirtualMachine.
+func (r *Reconciler) adoptOrphan(
+	ctx *pkgctx.VirtualMachineReplicaSetContext,
+	rs *vmopv1.VirtualMachineReplicaSet,
+	vm *vmopv1.VirtualMachine) error {
+
+	_, err := controllerutil.CreateOrPatch(ctx, r.Client, vm, func() error {
+		return controllerutil.SetControllerReference(rs, vm, r.Scheme())
```

Comment:
I wouldn't think we want to possibility of the creating a VM when this is just suppose to adopt

Reply:
I agree that it is confusing.  However, I used this because:
- the list of VMs passed down to this method are VMs that have been retrieved from the API server (using the `List` API call).  So, they already exist?
- It is _slightly_ cleaner then using `mergeFrom` and `cilent.Patch`.

LMK what you think.

Reply:
It could have been deleted between the List() and here (or even slightly before that if the ctr-runtime Client cache was stale). I don't think Patch() is too clumsy here but could always roll your own smaller wrapper, or just use Update() and if the RV is different just deal with the VM on a subsequent reconcile. 

Looks like there is another use of CreateOrPatch a little later on that looks problematic.

Reply:
Reverted to using Patch.